File: desktop/core/ai_converter.py

from groq import Groq, AuthenticationError, RateLimitError, APIConnectionError
from config.settings import load_config

SYSTEM_PROMPT = """You are a specialized Banglish text converter for Bangladeshi users.

DEFINITION:
Banglish = Bengali language written in English/Roman script, mixed with actual English words.
This is exactly how Bangladeshis type in social media, WhatsApp, Messenger, etc.

YOUR TASK:
Convert the given Banglish text to the requested language (Bangla Unicode or English).

STRICT RULES:
1. Preserve original meaning, tone, and emotion exactly
2. Handle informal language, slang, and code-mixing naturally
3. For bangla target: Use proper Unicode Bengali script only
4. For english target: Write natural, grammatically correct English
5. Return ONLY the converted text — zero explanations, zero quotes, zero extra text

EXAMPLES:
Input: "ami aj office theke fire khub tired, help lagbe"
Bangla result:  "আমি আজ অফিস থেকে ফিরে খুব ক্লান্ত, সাহায্য দরকার"
English result: "I came back from the office very tired today, I need help"

Input: "bhai ki korcho, call daw ektu urgent"
Bangla result:  "ভাই কী করছ, একটু কল দাও আর্জেন্ট"
English result: "Bro what are you doing, give me a call urgently"
"""


def convert_banglish(text: str, target: str = "bangla") -> str:

    config  = load_config()
    api_key = config.get("api_key", "").strip()

    if not api_key:
        return "❌ Groq API Key not set! Right-click tray icon → Settings"

    if not text or not text.strip():
        return ""

    if target == "bangla":
        instruction = "বাংলা (Bengali Unicode script)"
    else:
        instruction = "English (natural, grammatically correct)"

    user_prompt = f"Convert this Banglish text to {instruction}:\n\n{text.strip()}"

    try:
        client = Groq(api_key=api_key)

        response = client.chat.completions.create(
            model=config.get("model", "llama-3.3-70b-versatile"),
            max_tokens=1000,
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user",   "content": user_prompt},
            ]
        )

        return response.choices[0].message.content.strip()

    except AuthenticationError:
        return "❌ Invalid Groq API Key! Open Settings and check your key."
    except RateLimitError:
        return "❌ Rate limit reached. Wait a moment and try again."
    except APIConnectionError:
        return "❌ No internet connection."
    except Exception as e:
        return f"❌ Error: {str(e)[:80]}"

File: AIBanglish_KeyboardApp/app/src/main/java/com/rafiul/ai_banglish_keyboardapp/AIConverter.kt
package com.rafiul.ai_banglish_keyboardapp

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class AIConverter(private val context: Context) {

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    companion object {
        private const val API_URL = "https://api.groq.com/openai/v1/chat/completions"
        private const val MODEL   = "llama-3.3-70b-versatile"
        private const val SYSTEM  = """You are a Banglish converter for Bangladeshi users.
Banglish = Bengali written in Roman/English script mixed with English words.
RULES:
- Preserve meaning and tone exactly
- For bangla: return proper Unicode Bengali ONLY
- For english: return natural grammatically correct English
- Return ONLY converted text, nothing else
EXAMPLES:
"ami tired, ghume jabo" → বাংলা: "আমি ক্লান্ত, ঘুমাতে যাবো"
"bhai kemon acho" → English: "Bro how are you"
"""
    }

    suspend fun convert(text: String, toBangla: Boolean): Result<String> {
        return withContext(Dispatchers.IO) {
            try {
                val apiKey = getApiKey()
                if (apiKey.isEmpty()) return@withContext Result.failure(
                    Exception("API Key not set!\nGo to Settings → enter your Groq key.")
                )

                val target  = if (toBangla) "বাংলা (Bengali Unicode)" else "English"
                val userMsg = "Convert to $target:\n\n${text.trim()}"

                val body = JSONObject().apply {
                    put("model", MODEL)
                    put("max_tokens", 1000)
                    put("messages", JSONArray().apply {
                        put(JSONObject().apply { put("role","system"); put("content", SYSTEM) })
                        put(JSONObject().apply { put("role","user");   put("content", userMsg) })
                    })
                }.toString()

                val req = Request.Builder()
                    .url(API_URL)
                    .addHeader("Authorization", "Bearer $apiKey")
                    .addHeader("Content-Type", "application/json")
                    .post(body.toRequestBody("application/json".toMediaType()))
                    .build()

                val res = client.newCall(req).execute()
                if (!res.isSuccessful) return@withContext Result.failure(
                    Exception(when(res.code) {
                        401  -> "Invalid API key. Go to Settings and check."
                        429  -> "Rate limit reached. Wait a moment."
                        else -> "API error ${res.code}"
                    })
                )

                val text2 = JSONObject(res.body!!.string())
                    .getJSONArray("choices")
                    .getJSONObject(0)
                    .getJSONObject("message")
                    .getString("content").trim()

                Result.success(text2)

            } catch (e: Exception) {
                Result.failure(Exception("Error: ${e.message?.take(60)}"))
            }
        }
    }

    private fun getApiKey() = context
        .getSharedPreferences("banglish_prefs", Context.MODE_PRIVATE)
        .getString("api_key", "") ?: ""
}
